### PR TITLE
Make Parallel::printf atomic

### LINKED ISSUE #5722: Parallel::printf is not atomic
The deadlock output from `PrintElementInfo` is written by `Parallel::printf("%s", ss.str())`, but output from different elements is interleaved.  I don't know if the elements were on separate nodes.

Here's an example where we have boundary history information followed by the second half of the output of a `TimeStepId`:
```
   Key: (+2, [B19,(L2I2,L2I0,L1I0)]), history:
    Local Data:
     Time: 0:Slab[0.0000000000000000e+00,1.0000000000000001e-01]:0/1:0:0.0000000000000000e+00 (order 2)
     Time: 0:Slab[0.0000000000000000e+00,1.0000000000000001e-01]:0/1:1:1.9531250000000001e-04
     Time: 0:Slab[0.0000000000000000e+00,1.0000000000000001e-01]:1/16384:0:6.1035156250000003e-06 (order 2)
    Remote Data:
     Time: 0:Slab[0.0000000000000000e+00,1.0000000000000001e-01]:0/1:0:0.0000000000000000e+00 (order 18446744073709551615)
     Time: 0:Slab[0.0000000000000000e+00,1.0000000000000001e-01]:0/1:1:1.9531250000000001e-04
1.0000000000000001e-01]:1/512:0
    Key: (+1, [B8,(L2I0,L2I3,L1I0)]), next time: 0:Slab[0.0000000000000000e+00,1.0000000000000001e-01]:1/512:0
    Key: (-1, [B8,(L2I0,L2I1,L1I0)]), next time: 0:Slab[0.0000000000000000e+00,1.0000000000000001e-01]:1/512:0
    Key: (+2, [B8,(L2I0,L2I2,L1I1)]), next time: 0:Slab[0.0000000000000000e+00,1.0000000000000001e-01]:1/512:0
    Key: (+0, [B8,(L2I1,L2I2,L1I0)]), next time: 0:Slab[0.0000000000000000e+00,1.0000000000000001e-01]:1/512:0
```

### PR DESCRIPTION
**Pull request by wthrowe**
### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### CODE REVIEW

File: src/Parallel/Printf/Printf.hpp
Comment (nilsdeppe) [inline src/Parallel/Printf/Printf.hpp:162]: We could probably add an `fprint` that also takes a `std::string` as an argument that then prints to a text file. This will be useful for logging

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
LGTM. If this is annoying, just let us know :)

Reviewed with @wthrowe and @kidder

File: src/Parallel/Printf/Printf.hpp
Comment (knelli2) [reply src/Parallel/Printf/Printf.hpp:162]: If it's very simple to add, then I'd say do it. But if it's more than a few lines then maybe do it in a follow up?

**Review by knelli2 — CHANGES_REQUESTED** [review]
Needs a rebase

**Review by kidder — DISMISSED** [review]

**Review by nilsdeppe — DISMISSED** [review]
Amazing! Thank you :D

**Comment by knelli2** [comment]
We can ignore the clang-tidy errors in the generated code, but what about in `Printf.cpp`?

**Comment by wthrowe** [comment]
Added a new commit with the threadsafe error handling, and a fixup using that and hopefully fixing the macOS linking.

File: src/Utilities/ErrorHandling/Strerror.cpp
Comment (nilsdeppe) [inline src/Utilities/ErrorHandling/Strerror.cpp:14]: Could you add a comment (when you squash) about why 1000 is enough? I'm worried about overflow, but that worry might be misguided.

**Review by nilsdeppe — COMMENTED** [review]
LGTM. One small suggestion for when you squash.

**Review by knelli2 — COMMENTED** [review]
No additional comments. Also no need to wait for my approval to merge

File: src/Utilities/ErrorHandling/Strerror.cpp
Comment (wthrowe) [reply src/Utilities/ErrorHandling/Strerror.cpp:14]:
```diff
--- a/src/Utilities/ErrorHandling/Strerror.cpp
+++ b/src/Utilities/ErrorHandling/Strerror.cpp
@@ -9,6 +9,8 @@
 #include "Utilities/ErrorHandling/StrerrorWrapper.h"
 
 std::string strerror_threadsafe(const int errnum) {
+  // The documentation gives no guidance on how much space to reserve,
+  // but a typical English result is "No such file or directory".
   std::array<char, 1000> message{};
   if (spectre_strerror_r(errnum, message.data(), message.size()) != 0) {
     return "strerror failed";
```

**Review by wthrowe — COMMENTED** [review]
Also added a dependency of the fprintf test on Utilitites that I forgot to save before committing.

**Review by kidder — APPROVED** [review]

**Comment by kidder** [comment]
is this ready to merge?

**Comment by knelli2** [comment]
I have nothing else. @nilsdeppe ?

**Comment by nilsdeppe** [comment]
No, but @knelli2 says "Requested Changes"

**Review by knelli2 — APPROVED** [review]

**Pull request merged**
